packages/react-storage/src/components/StorageBrowser/actions/createEnhancedListHandler.ts

```diff
@@ -0,0 +1,92 @@
+import { AsyncDataAction } from '@aws-amplify/ui-react-core';
+import {
+  ListHandler,
+  ListHandlerOptions,
+  ListHandlerInput,
+  ListHandlerOutput,
+} from './types';
+
+interface SearchOptions<K> {
+  query: string;
+  filterKey: keyof K;
+}
+
+export interface EnhancedListHandlerOptions<K> extends ListHandlerOptions {
+  refresh?: boolean;
+  reset?: boolean;
+  search?: SearchOptions<K>;
+}
+
+interface EnhancedListHandler<T, K>
+  extends AsyncDataAction<
+    ListHandlerOutput<K>,
+    ListHandlerInput<EnhancedListHandlerOptions<K> & T>
+  > {}
+
+export const SEARCH_LIMIT = 10000;
+export const SEARCH_PAGE_SIZE = 1000;
+
+export const createEnhancedListHandler = <
+  T extends ListHandlerOptions<I>,
+  K,
+  I = never,
+>(
+  action: ListHandler<ListHandlerInput<T>, ListHandlerOutput<K>>
+): EnhancedListHandler<T, K> => {
+  return async function listActionHandler(prevState, { options, ...input }) {
+    const {
+      nextToken: _nextToken,
+      refresh,
+      reset,
+      search,
+      ...rest
+    } = options ?? {};
+
+    if (reset) {
+      return { items: [], nextToken: undefined };
+    }
+
+    // collect and filter results on `search`
+    if (search) {
+      const { query, filterKey } = search;
+
+      const result = [];
+      let nextNextToken = _nextToken;
+      do {
+        const output = await action({
+          ...input,
+          options: {
+            ...rest,
+            pageSize: SEARCH_PAGE_SIZE,
```

Comment:
should search page size be `Math.min(SEARCH_LIMIT-result.length, SEARCH_PAGE_SIZE)`?

Reply:
I considered making the SEARCH_LIMIT strict, but I think we should try to fetch as many items as possible per request for search, even if in some cases we end up fetching more than 10K items.

Reply:
Yeah I did not elaborate enough in my original comment, but you got my point perfectly. It sounds good to me!